### PR TITLE
Metrics exposed by PooledByteBufAllocator needs to be correctly synch…

### PR DESCRIPTION
**Pull request by normanmaurer**
…ronized

Motivation:

As we may access the metrics exposed of PooledByteBufAllocator from another thread then the allocations happen we need to ensure we synchronize on the PoolArena to ensure correct visibility.

Modifications:

Synchronize on the PoolArena to ensure correct visibility.

Result:

Fix multi-thread issues on the metrics

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch PTAL ... noticed this today. Please note we holds the lock on the PoolArena for the methods where things are updated already.

File: buffer/src/main/java/io/netty/buffer/PoolSubpage.java [hunk 1]
```diff
@@ -200,27 +200,35 @@ private long toHandle(int bitmapIdx) {
 
     @Override
     public String toString() {
-        if (!doNotDestroy) {
-            return "(" + memoryMapIdx + ": not in use)";
-        }
+        synchronized (chunk.arena) {
```
Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/PoolSubpage.java]: consider building the string outside the synchronized block.

File: buffer/src/main/java/io/netty/buffer/PoolChunkList.java [hunk 2]
```diff
@@ -190,37 +192,41 @@ private static int minUsage0(int value) {
...
 
-        StringBuilder buf = new StringBuilder();
-        for (PoolChunk<T> cur = head;;) {
-            buf.append(cur);
-            cur = cur.next;
-            if (cur == null) {
-                break;
+            StringBuilder buf = new StringBuilder();
+            for (PoolChunk<T> cur = head;;) {
+                buf.append(cur);
```
Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/PoolChunkList.java]: consider building the string outside the synchronized block. Trade off between potentially inconsistent state in the string and reducing the critical area.

Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolChunkList.java]: let me move the creation of the StringBuilder and the toString() call outside. The rest can stay inside as the PoolChunk itself will also need to hold the synchronized lock anyway.

File: buffer/src/main/java/io/netty/buffer/PoolSubpage.java (hunk 1 above)
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolSubpage.java]: +1

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (a7fe6c0) and 4.0 (d54ed19)

**Pull request closed without merging**